### ISSUE TITLE
SecurityException in Play dev environment

### ISSUE THREAD

**Issue by T-PWK** [playframework/playframework#8167]
I am getting an error as follows when in controller calling `play.data.FormFactory.form(FormData.class).bindFromRequest()` and when a form bean has any `play.data.validation.Constraints` annotations. 

The problem happens **only in development environment** when running application via `sbt run`. When I pack an application using `sbt stage` and run using generated bash, the error does not appear - all works as expected. 

Have you seen similar behaviour? Are you aware of any settings for a dev environment which could cause that?

**Edit**
This issue happens only when FormFactory.form(Class).bindFromRequest() is called directly from a controller handler method. If it is called using `CompletableFuture.supplyAsync(handler, ec)` where `ec` is either `HttpExecutionContext.current` or other extension of `CustomExecutionContext` then the problem does not happen.

```
java.util.concurrent.CompletionException: java.lang.SecurityException: setContextClassLoader
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:604)
	at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:614)
	at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:1983)
	at controllers.TriggerController.trigger(TriggerController.java:73)
	at admin_api_v1.Routes$$anonfun$routes$1.$anonfun$applyOrElse$184(Routes.scala:2594)
	at play.core.routing.HandlerInvokerFactory$$anon$4.resultCall(HandlerInvoker.scala:137)
	at play.core.routing.HandlerInvokerFactory$JavaActionInvokerFactory$$anon$8$$anon$2$$anon$1.invocation(HandlerInvoker.scala:108)
	at play.core.j.JavaAction$$anon$1.call(JavaAction.scala:88)
Caused by: java.lang.SecurityException: setContextClassLoader
	at java.util.concurrent.ForkJoinWorkerThread$InnocuousForkJoinWorkerThread.setContextClassLoader(ForkJoinWorkerThread.java:244)
	at play.api.inject.ContextClassLoaderInjector.withContext(Injector.scala:125)
	at play.api.inject.ContextClassLoaderInjector.instanceOf(Injector.scala:118)
	at play.inject.DelegateInjector.instanceOf(DelegateInjector.java:22)
	at play.data.validation.DefaultConstraintValidatorFactory.getInstance(DefaultConstraintValidatorFactory.java:26)
	at org.hibernate.validator.internal.engine.constraintvalidation.ConstraintValidatorManager.createAndInitializeValidator(ConstraintValidatorManager.java:145)
	at org.hibernate.validator.internal.engine.constraintvalidation.ConstraintValidatorManager.getInitializedValidator(ConstraintValidatorManager.java:109)
	at org.hibernate.validator.internal.engine.constraintvalidation.ConstraintTree.getConstraintValidatorNoUnwrapping(ConstraintTree.java:301)
	at org.hibernate.validator.internal.engine.constraintvalidation.ConstraintTree.getConstraintValidatorInstanceForAutomaticUnwrapping(ConstraintTree.java:242)
	at org.hibernate.validator.internal.engine.constraintvalidation.ConstraintTree.getInitializedConstraintValidator(ConstraintTree.java:163)
```

### JDK: Oracle 1.8.0_161

java version "1.8.0_161"
Java(TM) SE Runtime Environment (build 1.8.0_161-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.161-b12, mixed mode)

**Comment by wsargent** [comment]
Seems like it might need an explicit executor:

https://stackoverflow.com/questions/21440563/webstart-application-securityexception-setcontextclassloader

**Comment by patrickliangli** [comment]
I also ran into this issue, but only when I use Deadbolt to secure the controller method. If I remove it, the problem goes away.

**Comment by marcospereira** [comment]
Related to #7751.

**Comment by mkurz** [comment]
@patrickliangli Can you give more details? Which version of Play are you using? And also which version of deadbolt? Which deadbolt annotation did you use (I maintain deadbolt so this information would help)?

**Comment by mkurz** [comment]
I am pretty sure this issue here, as well as #7751 and #7615 (which has been fixed with another workaround meanwhile) **will be fixed automagically with upcoming Play 2.8**.

Why?
Because Play 2.8 will [switch to Akka 2.6](9236)! :laughing:
As it turns out, up until Akka 2.5 (and therefore up to Play 2.7), Akka uses [it's own implementation](https://github.com/akka/akka/pull/22560) of [`ForkJoinPool`](https://github.com/akka/akka/blob/v2.5.22/akka-actor/src/main/java/akka/dispatch/forkjoin/ForkJoinPool.java), instead of the one that comes with the JDK - because the Akka Team [wasn't satisfied](https://github.com/akka/akka/issues/21910) with its performance (turns out before that Akka was using the FJP from Scala, but Scala 2.12 switched to the JDK one... but that doesn't matter for us here). Anyway, the plan for upcoming Akka 2.6 is to remove it's own embedded FJP: https://github.com/akka/akka/issues/26180 - and if this plan becomes reality this and the other issues would be fixed.
But why would removing the embedded FJP from Akka fix all those issues in Play?
Read [this conversation](https://github.com/akka/akka/issues/23662) I once had with the Akka team: Right now, when methods like eg. `parallelStream` (or similiar) try to find out in which ThreadPool they should run their operations, they call `ForkJoinTask#inForkJoinPool()` to make that decision. That `inForkJoinPool` returns `false` for Akka's embedded FJP implementation (even though it also is a FJP...), meaning such methods always choose the default commonPool FJP as fallback. However, as soon as Akka 2.6 switches to the JDK's FJP implementation, `inForkJoinPool` will then return `true`, and any `parallel` or any `thenxxxAsync` method (which does not explicitly provide an execution context) will then not use the common pool anymore, but the one that akka provides.

**Comment by marcospereira** [comment]
It sounds like good news, @mkurz.

I would say that on our side, it could be useful to have a test to confirm this is not happening for 2.8+ and avoid regressions. While we don't have that, let's keep this issue open.

Best.

**Comment by mkurz** [comment]
@marcospereira https://github.com/akka/akka/issues/26180 has been fixed (https://github.com/akka/akka/pull/26893). As soon as we upgrade to akka 2.6.0-M2 in master we should be able to write a test that confirms this is fixed. I will assign myself.

**Comment by mkurz** [comment]
I finally found time to write a test to confirm `SecurityException` will not be thrown anymore when using `map`, `filter`, etc. on a **parallel** stream within a Play action method. See #9594

However, when using `CompletableFuture.supplyAsync(...)` you still have to pass an explicit ec, because Java uses the common thread pool for this method by default. There is nothing Play can do about this (it's no an issue with Play actually).

I suggest closing this issue as soon as #9594 got merged.
